### ISSUE TITLE
[Question]: Slowness of LLM inference step using chat engine

### ISSUE THREAD

**Issue by Swarnashree** [run-llama/llama_index#8051]
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

Hi, I am trying to setup a basic RAG chat pipeline with the below code and I am using [Mistral 7B instruct model](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1). I am noticing the generate function under the hood is a lot slower than the generate function from [liltom-eth/llama2-webui](https://github.com/liltom-eth/llama2-webui/blob/379273b408462ed04cdb8622f7f9fa080fdfff6c/llama2_wrapper/model.py#L251). (all other steps of retrieving nodes to be used in the prompt etc are fast enough. Only the step of generating content from the LLM is slow)

I made sure to have the same settings for the LLM on both the webui project and also the llama-index project by setting these in the llama-index project:
```python
model_kwargs={"torch_dtype": torch.float16},
generate_kwargs={"do_sample": True, "top_p": 0.95, "top_k": 50, "temperature": 0.8, "num_beams":1}
```

Llama-index RAG pipeline is as follows:
```python
from llama_index.llms import HuggingFaceLLM
from llama_index import LangchainEmbedding
from llama_index.langchain_helpers.text_splitter import SentenceSplitter
from langchain.embeddings import HuggingFaceEmbeddings

llm = HuggingFaceLLM(
    system_prompt="Consider you are a smart finanical assistant who has the capability to interpret SEC filings and answer the questions for the given text as expected.",
    tokenizer_name="mistralai/Mistral-7B-Instruct-v0.1",
    model_name="mistralai/Mistral-7B-Instruct-v0.1",
    device_map="auto",
   model_kwargs={"torch_dtype": torch.float16},
   generate_kwargs={"do_sample": True, "top_p": 0.95, "top_k": 50, "temperature": 0.8, "num_beams":1}
)

prompt_helper = PromptHelper.from_llm_metadata(llm_metadata=llm.metadata)
embed_model = LangchainEmbedding(langchain_embeddings=HuggingFaceEmbeddings(
        model_name="sentence-transformers/all-mpnet-base-v2",
    model_kwargs={'device': 'cuda:1'},
    encode_kwargs={'normalize_embeddings': False}
    ))
service_context = ServiceContext.from_defaults(
    llm=llm,
    prompt_helper=prompt_helper,
    node_parser=SimpleNodeParser(text_splitter=SentenceSplitter()),
    embed_model=embed_model)

index = VectorStoreIndex.from_documents(documents, service_context=service_context)
chat_engine = index.as_chat_engine(chat_mode='context',retriever_mode='embedding', similarity_top_k=2)
        
stream_response=chat_engine.stream_chat("what are the tax implications of the said filing?")
stream_response.print_response_stream()
```

Any help on where the speed difference is coming, would be appreciated
I am running them on 4 T4 GPUs and also noticed the GPU utilization during the generate function is much higher % than that in the llama-index project which makes me wonder if there is some inefficiency of GPU utilization with the way I am setting up the RAG pipeline

**Comment by Swarnashree** [comment]
@dosu-bot just to be super clear, I meant the generate function [here](https://github.com/run-llama/llama_index/blob/85de3d9a503fec159962569fb77329a43af4594e/llama_index/llms/huggingface.py#L254) and [llama2-ui](https://github.com/liltom-eth/llama2-webui/blob/379273b408462ed04cdb8622f7f9fa080fdfff6c/llama2_wrapper/model.py#L251). 
As you can see, it is not related to the LLMQuestionGenerator or OpenAIQuestionGenerator classes.

> For example, the LlamaIndex project might be using a different version of PyTorch or a different CUDA toolkit version, which could affect GPU utilization.

both were run using the same hardware setup and conda environment i.e. so same package versions.

**Comment by logan-markewich** [comment]
@Swarnashree I think it's likely slow because we are packing the context window to near its limit.

If you just prompt the llm with `llm.complete("Hello!")` it will return very fast.

But if you fill the context window with something like 3900 tokens, inference will be slower

**Comment by Swarnashree** [comment]
> @Swarnashree I think it's likely slow because we are packing the context window to near its limit.
> 
> If you just prompt the llm with `llm.complete("Hello!")` it will return very fast.
> 
> But if you fill the context window with something like 3900 tokens, inference will be slower

Yes, I agree, just to mention, I copied the final RAG based prompt that was generated by llama-index project onto the llama2-ui project. So both have the same incoming prompt into the llm

**Comment by logan-markewich** [comment]
If you spot something that could make it faster in our source code, I encourage a PR!
https://github.com/run-llama/llama_index/blob/main/llama_index/llms/huggingface.py

**Comment by logan-markewich** [comment]
> Yes, I agree, just to mention, I copied the final RAG based prompt that was generated by llama-index project onto the llama2-ui project. So both have the same incoming prompt into the llm

Interesting. If you have time to debug this it would be appreciated, off-hand I'm not sure what the difference is

**Comment by Swarnashree** [comment]
> Interesting. If you have time to debug this it would be appreciated, off-hand I'm not sure what the difference is

Hi @logan-markewich, after spending some time debugging, I see that the bottleneck comes from starting the thread for `write_response_history` in context.py [code](https://github.com/run-llama/llama_index/blob/06127ec09966e8df2fcd4f03a1b53ec566b4a43d/llama_index/chat_engine/context.py#L200)

To dig further, I tried all combinations of using/not using background(bg) threads: 
i.e. `self._model.generate` thread in huggingface.py [code]()(lets call it func1) and `write_response_history` in context.py (lets call it func2) and bottleneck is evident ONLY when both are on bg threads

func1 on bg thread, func2 not on bg thread: speed of inference is as expected (:+1: )
func1 on bg thread, func2 not on bg thread: speed is not as expected (:-1: )
func1 not on bg thread, func2 on bg thread: speed of inference is as expected (:+1: )
func1 not on bg thread, func2 not on bg thread: speed of inference is as expected (:+1: )

This probably means the GPU performance is being degraded due to the over multi-threading. Is there any way to set this as perhaps a configuration to limit the number of threads?

**Comment by logan-markewich** [comment]
@Swarnashree amazing debugging!

I'm not 100% sure on how best to resolve this 🤔 the thread is needed in order for streaming to work.

I'll look into this and chat with the team, and see if we can figure out a fix.
